### PR TITLE
Add support for Enclustra Mercury XU* modules

### PR DESCRIPTION
**Pull request by kgugala**
This PR adds suport for the Cortex R5 CPU in the Xilinx Zynq MP SoC on Enclustra's Mercury XU* Modules.

It also adds possibility to disable ECC checking in the Cortex R Tightly Coupled Memory (TCM)

### CODE REVIEW

**Comment by kgugala** [comment]
ping @galak @ioannisg

**Comment by kgugala** [comment]
This PR is hanging here for quite some time and it is blocking our next PRs adding AMP support for Cortexes-R. I have rebased the PR onto current master. @galak @ioannisg can you take a look on this one, or point somebody else who can review this code? Thanks in advance

File: arch/arm/core/cortex_r/Kconfig [hunk 1]
```diff
@@ -86,6 +86,12 @@ config GEN_ISR_TABLES
 config GEN_IRQ_VECTOR_TABLE
 	default n
 
+config DISABLE_TCM_ECC
+	bool
+	default n
```
Comment (ioannisg) [inline arch/arm/core/cortex_r/Kconfig:91]: This option is introduced here, right?
Any reason this is hidden?
Also, default n is not required for boolean Kconfigs

File: arch/arm/core/cortex_r/reset.S [hunk 2]
```diff
@@ -99,4 +100,8 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 	bl z_arm_watchdog_init
 #endif
 
+#if defined(CONFIG_DISABLE_TCM_ECC)
+	bl disable_ecc
```
Comment (ioannisg) [inline arch/arm/core/cortex_r/reset.S:104]: Is this something that is SoC specific or ARCH specific?
If SoC specific, I would suggesting doing this under PLATFORM_SPECIFIC_INIT (see for example Cortex-M reset.S)

File: boards/arm/mercury_xu/mercury_xu_board_defconfig [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2019, Antmicro
+CONFIG_ARM=y
+CONFIG_SOC_XILINX_ZYNQMP=y
+CONFIG_BOARD_MERCURY_XU_BOARD=y
+
+CONFIG_ISR_STACK_SIZE=1024
+CONFIG_THREAD_STACK_INFO=y
```
Comment (ioannisg) [inline boards/arm/mercury_xu/mercury_xu_board_defconfig:7]: Why was this needed here?

**Comment by stephanosio** [comment]
This PR looks ancient. Is this still being worked on?

File: arch/arm/core/cortex_r/reset.S (hunk 2 above)
Comment (stephanosio) [reply arch/arm/core/cortex_r/reset.S:104]: This should be at arch-level. TCM ECC is a feature that is directly integrated into the Cortex-R processors.

File: arch/arm/core/cortex_r/tcm.c [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+void disable_ecc(void)
+{
+	__asm__(
+		"mrc p15, 0, r0, c1, c0, 1\n\t" /* read ACTLR */
+		"bic r0, r0, #0xe000000\n\t"  /* disable ecc */
+		"mcr p15, 0, r0, c1, c0, 1\n\t" /* write ACTLR */
+	);
```
Comment (stephanosio) [inline arch/arm/core/cortex_r/tcm.c:13]: This can make use of the `__{get,set}_ACTLR` functions.

https://github.com/zephyrproject-rtos/cmsis/blob/a2b82e79bd8ac156fb283d5f52fa0a2373962cf9/CMSIS/Core_R/Include/cmsis_cp15.h#L35-L51

**Review by stephanosio — CHANGES_REQUESTED** [review]
Needs rebase

File: boards/arm/mercury_xu/Kconfig.board [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2019, Antmicro
+# SPDX-License-Identifier: Apache-2.0
+config BOARD_MERCURY_XU_BOARD
+	bool "Mercury XU Board"
+	depends on SOC_XILINX_ZYNQMP
+	select DISABLE_TCM_ECC
```
Comment (stephanosio) [inline boards/arm/mercury_xu/Kconfig.board:6]: Unless there is a board-level reason as to why `DISABLE_TCM_ECC` must be disabled, it should be specified in the `defconfig` (and yes, the symbol should have a prompt as @ioannisg pointed out).

File: boards/arm/mercury_xu/board.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2019, Antmicro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef __INC_BOARD_H
```
Comment (stephanosio) [inline boards/arm/mercury_xu/board.h:6]:
```suggestion
#ifndef ZEPHYR_BOARDS_ARM_MERCURY_XU_BOARD_H_
```

File: boards/arm/mercury_xu/mercury_xu_board.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2019, Antmicro
+identifier: mercury_xu_board
+name: MERCURY-XU-BOARD
+type: mcu
+arch: arm
+toolchain:
+  - zephyr
+  - gccarmemb
```
Comment (stephanosio) [inline boards/arm/mercury_xu/mercury_xu_board.yaml:8]: `gccarmemb` is obsolete.

```suggestion
  - gnuarmemb
```

File: boards/arm/mercury_xu/mercury_xu_board.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+ram: 64
+testing:
+  ignore_tags:
+   - nmi
+   - benchmark
+   - interrupt
+   - bluetooth
+   - socket
+   - crypto
+   - net
```
Comment (stephanosio) [inline boards/arm/mercury_xu/mercury_xu_board.yaml:18]: These are probably no longer necessary.

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by kgugala** [comment]
@stephanosio @ioannisg  thanks for the comments. We'll rebase this and test again

**Comment by kgugala** [comment]
@stephanosio @ioannisg I have rebased the PR. I believe the comments are addressed. Can you take a look on this one?

File: arch/arm/include/aarch32/cortex_r/tcm.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
+/*
```
Comment (stephanosio) [inline arch/arm/include/aarch32/cortex_r/tcm.h:1]: Please add this file to `arch/arm/include/aarch32/cortex_a_r` directory (`cortex_r` directory will be removed in future releases).

File: arch/arm/core/aarch32/cortex_a_r/tcm.c [hunk 10]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2019, Antmicro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <cmsis_cp15.h>
```
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_a_r/tcm.c:7]: Please include `arch/arm/aarch32/cortex_a_r/cmsis.h`

File: arch/arm/core/aarch32/cortex_a_r/tcm.c [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+void z_arm_tcm_disable_ecc(void)
+{
+	uint32_t actlr;
+
+	actlr = __get_ACTLR();
+	actlr &= ~0xe000000;
```
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_a_r/tcm.c:14]: Given that these bits are defined in the CMSIS, please make use of the definitions:
https://github.com/zephyrproject-rtos/cmsis/blob/a2b82e79bd8ac156fb283d5f52fa0a2373962cf9/CMSIS/Core_R/Include/core_cr.h#L511-L518

```suggestion
	actlr &= ~(ACTLR_ATCMPCEN_Msk | ACTLR_B0TCMPCEN_Msk | ACTLR_B1TCMPCEN_Msk);
```

File: arch/arm/include/aarch32/cortex_r/tcm.h (hunk 9 above)
Comment (kgugala) [reply arch/arm/include/aarch32/cortex_r/tcm.h]: done

File: arch/arm/core/aarch32/cortex_a_r/tcm.c (hunk 10 above)
Comment (kgugala) [reply arch/arm/core/aarch32/cortex_a_r/tcm.c]: done

File: arch/arm/core/aarch32/cortex_a_r/tcm.c (hunk 11 above)
Comment (kgugala) [reply arch/arm/core/aarch32/cortex_a_r/tcm.c]: done

**Comment by kgugala** [comment]
I have rebased the PR on top of current master. @stephanosio can you check if your comments have been resolved?

File: arch/arm/include/aarch32/cortex_a_r/tcm.h [hunk 12]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2019, Antmicro
```
Comment (PiotrZierhoffer) [inline arch/arm/include/aarch32/cortex_a_r/tcm.h:2]: I believe these should all be 2020 now

File: boards/arm/mercury_xu/Kconfig.defconfig [hunk 13]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2019, Antmicro
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_MERCURY_XU
+
+config BOARD
+	default "mercury_xu"
+
+config BUILD_OUTPUT_STRIPPED
```
Comment (PiotrZierhoffer) [inline boards/arm/mercury_xu/Kconfig.defconfig:9]: Is this needed? I see this setting in 3 x86 boards, but I don't think it should be a default

**Review by PiotrZierhoffer — APPROVED** [review]
LGTM, minor comments only

File: arch/arm/include/aarch32/cortex_a_r/tcm.h (hunk 12 above)
Comment (kgugala) [reply arch/arm/include/aarch32/cortex_a_r/tcm.h]: should be fixed now (here and in all the other occurrences)

File: boards/arm/mercury_xu/Kconfig.defconfig (hunk 13 above)
Comment (kgugala) [reply boards/arm/mercury_xu/Kconfig.defconfig]: indeed, this is not needed. Removed

**Comment by kgugala** [comment]
@stephanosio can you re-review this PR? I believe all your comments were addressed

**Review by stephanosio — COMMENTED** [review]
Looks ok overall.

**Comment by wsipak** [comment]
@stephanosio Thank you. Your suggested changes are included in the code now.

File: arch/arm/include/aarch32/cortex_a_r/tcm.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ */
+void z_arm_tcm_disable_ecc(void);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* _ASMLANGUAGE */
+
+#endif /* ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_R_TCM_H_ */
```
[the remark this answers is not in the document]
Comment (wsipak) [reply arch/arm/include/aarch32/cortex_a_r/tcm.h]: Thank you. This should be ok now.

**Review by stephanosio — APPROVED** [review]
Looks ok to me.

p.s. it would be nice to have docs.
